pkg/reconciler/workload/namespace/namespace_reconcile_placementbind.go

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+
+	schedulingv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/scheduling/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+)
+
+// bindNamespaceReconciler checks the namespaces and binds the namespace to all matched placements
+// The state will be set to bound so scheduler will be started.
+// TODO this should be reconsidered when we want lazy binding.
+type bindNamespaceReconciler struct {
+	listPlacement func(clusterName logicalcluster.Name) ([]*schedulingv1alpha1.Placement, error)
+
+	patchNamespace func(ctx context.Context, clusterName logicalcluster.Name, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (*corev1.Namespace, error)
+}
+
+func (r *bindNamespaceReconciler) reconcile(ctx context.Context, ns *corev1.Namespace) (reconcileStatus, *corev1.Namespace, error) {
+	clusterName := logicalcluster.From(ns)
+
+	var currentPlacementAnnotation schedulingv1alpha1.PlacementAnnotation
+
+	placementsValue, foundPlacement := ns.Annotations[schedulingv1alpha1.PlacementAnnotationKey]
+	if foundPlacement {
+		if err := json.Unmarshal([]byte(placementsValue), &currentPlacementAnnotation); err != nil {
+			klog.Errorf("failed to unmarshal placement annotation %q of namespace %s|%s: %v", placementsValue, clusterName, ns.Name, err)
+			currentPlacementAnnotation = schedulingv1alpha1.PlacementAnnotation{}
+		}
+	}
+
+	validPlacements, err := r.validPlacements(clusterName, ns)
```

Comment:
does this mean this annotation is always added, even if there is no matching Placement?

Reply:
updated, it there is not valid placement, we will remove the annotation